### PR TITLE
Revert #849[Use protobufs for communication with apiserver] in v1.6

### PR DESCRIPTION
**Pull request by skmatti**
Fixes https://github.com/kubernetes/ingress-gce/issues/1049

### CODE REVIEW

File: cmd/glbc/app/clients.go [hunk 1]
```diff
@@ -44,18 +44,17 @@ const (
...
-// NewKubeConfig returns a Kubernetes client config given the command line settings.
-func NewKubeConfig() (*rest.Config, error) {
-	config, err := newKubeConfig()
-	if err != nil {
-		return nil, err
-	}
+// KubeConfigForProtobuf returns a shallow copy of client config after updating
+// content type to protobuf.
+func KubeConfigForProtobuf(kubeConfig *rest.Config) *rest.Config {
+	configShallowCopy := *kubeConfig
```
Comment (bowei) [inline cmd/glbc/app/clients.go]: why not deep copy?

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
...
 	ctx.BackendConfigInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			beConfig := obj.(*backendconfigv1beta1.BackendConfig)
+			beConfig, ok := obj.(*backendconfigv1beta1.BackendConfig)
+			if !ok {
+				klog.Errorf("Error decoding object(%T) into BackendConfig", obj)
+				return
+			}
 			ings := operator.Ingresses(ctx.Ingresses().List()).ReferencesBackendConfig(beConfig, operator.Services(ctx.Services().List())).AsList()
+			if len(ings) == 0 {
```
Comment (bowei) [inline pkg/controller/controller.go]: I think Enqueue already logs?

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
...
 		UpdateFunc: func(old, cur interface{}) {
 			if !reflect.DeepEqual(old, cur) {
-				beConfig := cur.(*backendconfigv1beta1.BackendConfig)
+				beConfig, ok := cur.(*backendconfigv1beta1.BackendConfig)
+				if !ok {
+					klog.Errorf("Error decoding object(%T) into BackendConfig", cur)
+					return
+				}
 				ings := operator.Ingresses(ctx.Ingresses().List()).ReferencesBackendConfig(beConfig, operator.Services(ctx.Services().List())).AsList()
+				if len(ings) == 0 {
```
Comment (bowei) [inline pkg/controller/controller.go]: Same here

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
 	// BackendConfig event handlers.
 	ctx.BackendConfigInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			beConfig := obj.(*backendconfigv1beta1.BackendConfig)
+			beConfig, ok := obj.(*backendconfigv1beta1.BackendConfig)
+			if !ok {
+				klog.Errorf("Error decoding object(%T) into BackendConfig", obj)
```
Comment (bowei) [inline pkg/controller/controller.go]: Let use standard terminology:

"Cannot cast obj to BackendConfig (obj type is %T)"

fix below as well

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
...
+				}
+				klog.V(3).Infof("BackendConfig %s/%s updated, enqueuing ingresses %v", beConfig.Namespace, beConfig.Name, ings)
 				lbc.ingQueue.Enqueue(convert(ings)...)
 			}
 		},
 		DeleteFunc: func(obj interface{}) {
-			beConfig := obj.(*backendconfigv1beta1.BackendConfig)
+			beConfig, ok := obj.(*backendconfigv1beta1.BackendConfig)
+			if !ok {
+				klog.V(3).Infof("Failed to decode object(%T) as BackendConfig, casting into DeletedFinalStateUnknown", obj)
```
Comment (bowei) [inline pkg/controller/controller.go]: see above

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
...
+				var tombstone cache.DeletedFinalStateUnknown
+				tombstone, ok = obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					klog.Errorf("Error decoding object(%T) into DeletedFinalStateUnknown", obj)
```
Comment (bowei) [inline pkg/controller/controller.go]: see above

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
...
+					return
+				}
+				if beConfig, ok = tombstone.Obj.(*backendconfigv1beta1.BackendConfig); !ok {
+					klog.Errorf("Error decoding deleted state(%T) of object into BackendConfig", tombstone.Obj)
```
Comment (bowei) [inline pkg/controller/controller.go]: see above

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -191,20 +191,56 @@ func NewLoadBalancerController(
...
+					return
+				}
+			}
 			ings := operator.Ingresses(ctx.Ingresses().List()).ReferencesBackendConfig(beConfig, operator.Services(ctx.Services().List())).AsList()
+			if len(ings) == 0 {
```
Comment (bowei) [inline pkg/controller/controller.go]: again, check that we log in Enqueue so we don't have to log here as well

File: cmd/glbc/app/clients.go (hunk 1 above)
Comment (skmatti) [reply cmd/glbc/app/clients.go]: We are modifying only the top level variable and this is used as read only by kubeclient for most part. I modified this to create separate configurations in each case.

File: pkg/controller/controller.go (hunk 2 above)
Comment (skmatti) [reply pkg/controller/controller.go]: Not when ingresses are not enqueued(empty slice). I preserved this log and removed the other.

Comment (bowei) [reply pkg/controller/controller.go]: can we make Enqueue log for empty slice and delete the extra logs outside the call

**Comment by freehan** [comment]
Discussed offline. @skmatti will rollback the protobuf PR for ingress 1.6 as for now.

File: pkg/controller/controller.go (hunk 2 above)
Comment (skmatti) [reply pkg/controller/controller.go]: Removed these logs. Not sure adding this log inside enqueue would be any help as this backendconfig obj is not available inside that method.

**Comment by skmatti** [comment]
> Discussed offline. @skmatti will rollback the protobuf PR for ingress 1.6 as for now.
> 
> Fix will be pushed to HEAD and cherrypick back to 1.6 1.7

Rolled back #849 in v1.6. I think this more cleaner as this change never got released in any GKE versions.

**Pull request merged**